### ISSUE TITLE
Event Properties with Callbacks

### ISSUE THREAD

**Issue by splashsky** [Annoraaq/phaser-grid-movement-plugin#48]
Hey again!

In the game I am creating, it will be necessary to be able to know when I am interacting with certain tiles on the map. Since the plugin is the one handling collision checks and movement, I suggest a number of properties that could be super useful for map events. I'll also give a use case for each event, but these mechanics would be implemented by the game developer. I am only suggesting we add properties for certain tilemap-based events.

- `onEnter` - this event would trigger any time a certain tile has been entered. This could be used in such cases as initiating fights, minigames or puzzle mechanics.

- `onStanding` - this event will trigger every frame that the character is standing on the given tile. This could be used for automated platforms for puzzles, weight buttons or healing pads.

- `onExit` - this event will trigger when the character leaves a given tile. This could be used to trigger trap doors, or other cinematic effects such as rockslides behind the player.

- `onFacing` - this event could be triggered when the player is adjacent to and facing the given tile. This could be used to initiate interactions with objects, dialog with NPCs, et cetera. This would, however, require adding a "facing" mechanic to the plugin itself to implement.

The uses for such events are immense; players could define "trigger" areas on a tilemap that would start certain events or cinematics, and it would facilitate a lot of dynamic interactions with the world such as dialogs, shops, puzzles, platforms, et cetera.

**Comment by splashsky** [comment]
To clarify implementation, I thought of two scenarios:

1. The property is a string with the name of the function to be called (or perhaps the name of the event for an event listener), and the developer would implement the function

2. The property is simply a bool (true/false) that would fire the name of the event for an event listener, and the developer would still implement the function

I'm not sure which is the more "proper" implementation, but either one would work fine imo. 1 is more flexible if the developer wants to use a signal name other than `onEnter`, for example.

**Comment by Annoraaq** [comment]
Hi splashsky,

first of all I want to thank you for your support and all the good suggestions. It helps me a lot to know which features are demanded most.

I think it is a good idea to have more events like this. I would rather offer observables than events or callbacks. (like the `movementStarted` observable for example).

What do you think of the following suggestions:

- Instead of `onEnter` or `onExit` there could be an observable `positionChanged` that provides two tile positions (1. the tile the player is leaving and 2. the tile the player moved to). This is more flexible and allows to construct `onEnter` and `onExit` observables and more.

- Instead of repeatedly firing events for `onStanding` or `onFacing` there could be methods that you can call on demand like `isCharStanding(charId: string): boolean` and `isCharFacing(charId: string, direction: Direction): boolean`. You can still check it on every update if you like.

**Comment by splashsky** [comment]
I like the idea of the observables. How would you recommend we tie tiles to certain behaviors? i.e. if I have a type of tile I want to always be a sign, or a shop. In some way, we need to streamline giving tiles behavior.

I am 100% on-board with the on-demand functions. 👍🏻

**Comment by Annoraaq** [comment]
I think I now see your case. One example would be to trigger some action (like opening a shop window) whenever a user enters a "shop" tile. Is that correct?

In that case you could subscribe to our fictional `positionChanged` observable like this:

```typescript
gridMovementPlugin.positionChanged().subscribe(
({characterId, exitTile, enterTile}) => {
  if (isShop(enterTile)) {
    openShopWindow();
  }
});
```

If you want to only listen to this for **one** character (for example only for the player and not for NPCs), you can also filter the observable:

```typescript
gridMovementPlugin.positionChanged()
  .pipe(filter(({characterId}) => characterId === 'player'))
  .subscribe(
    // ...
  );
```

Does that help?

**Comment by splashsky** [comment]
It looks excellent. I presume we could grab any tile property we wanted via this method, so it's flexible enough. Excellent work!

**Comment by Annoraaq** [comment]
Thank you. I will schedule the issue for the next milestone as well. You are bringing great ideas!

Btw: Whenever you want to share your project, please do so. I think it would be great to show a list of projects created with the plugin.

**Comment by Annoraaq** [comment]
@splashsky 
Published with 1.7.0 release.

There is now an observable `positionChanged()` as well as a method for getting movement state: `isMoving()`. 
I also added methods for getting (`getFacingDirection()`) and setting (`turnTowards()`) the facing direction.

There are three new examples to demonstrate their usage.
